api/server/sdk/cluster_pair.go

```diff
@@ -0,0 +1,135 @@
+/*
+Package sdk is the gRPC implementation of the SDK gRPC server
+Copyright 2018 Portworx
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package sdk
+
+import (
+	"context"
+
+	"github.com/libopenstorage/openstorage/api"
+	"github.com/libopenstorage/openstorage/cluster"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// ClusterPairServer is an implementation of the gRPC OpenStorageClusterServer interface
+type ClusterPairServer struct {
+	server *Server
+}
+
+func (s *ClusterPairServer) cluster() cluster.Cluster {
+	return s.server.cluster()
+}
+
+// Create a new cluster with remote pair
+func (s *ClusterPairServer) Create(
+	ctx context.Context,
+	req *api.SdkClusterPairCreateRequest,
+) (*api.SdkClusterPairCreateResponse, error) {
+
+	if req.GetRequest() == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "Must supply valid request")
+	}
+
+	resp, err := s.cluster().CreatePair(req.GetRequest())
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Cannot create cluster with remote pair %s : %v",
+			req.GetRequest().GetRemoteClusterIp(), err)
+	}
+
+	return &api.SdkClusterPairCreateResponse{
+		Result: resp,
+	}, nil
+}
+
+// Inspect information about a cluster pair
+func (s *ClusterPairServer) Inspect(
+	ctx context.Context,
+	req *api.SdkClusterPairInspectRequest,
+) (*api.SdkClusterPairInspectResponse, error) {
+
+	if len(req.GetId()) == 0 {
```

Comment:
Redundant check, should be done by the implementation

Reply:
That is ok, the message here is that this check returns a 4xx or codes.InvalidArgument error code. While returning errors from the golang interface would have to be a 5xx or codes.Internal since the SDK would not know.

To make it clear to the caller, this check provides a very specific message that the Id is required and the error is caused by an invalid request.

Reply:
And a clear message is already being returned by the implementation. It is not the SDKs job to verify this. Different implementations could have different requirements.

Reply:
We discussed offline, and we think that the underlying Golang interfaces can do one of two things:

* Return grpc code status errors
* Return wellknown error codes which we can then map to gRPC codes.